crates/winch/src/compiler.rs

```diff
@@ -1,50 +1,136 @@
 use anyhow::Result;
+use cranelift_codegen::{Final, MachBufferFinalized};
 use object::write::{Object, SymbolId};
 use std::any::Any;
+use std::sync::Mutex;
+use wasmparser::FuncValidatorAllocations;
+use wasmtime_cranelift_shared::obj::ModuleTextBuilder;
 use wasmtime_environ::{
-    CompileError, DefinedFuncIndex, FuncIndex, FunctionBodyData, FunctionLoc, ModuleTranslation,
-    ModuleTypes, PrimaryMap, Tunables, WasmFunctionInfo,
+    CompileError, DefinedFuncIndex, FilePos, FuncIndex, FunctionBodyData, FunctionLoc,
+    ModuleTranslation, ModuleTypes, PrimaryMap, Tunables, WasmFunctionInfo,
 };
 use winch_codegen::TargetIsa;
 
+use crate::environ::FuncEnv;
+
 pub(crate) struct Compiler {
     isa: Box<dyn TargetIsa>,
+    allocations: Mutex<Vec<FuncValidatorAllocations>>,
 }
 
+struct CompiledFunction(MachBufferFinalized<Final>);
+
 impl Compiler {
     pub fn new(isa: Box<dyn TargetIsa>) -> Self {
-        Self { isa }
+        Self {
+            isa,
+            allocations: Mutex::new(Vec::new()),
+        }
+    }
+
+    fn take_allocations(&self) -> FuncValidatorAllocations {
+        self.allocations
+            .lock()
+            .unwrap()
+            .pop()
+            .unwrap_or_else(Default::default)
+    }
+
+    fn save_allocations(&self, allocs: FuncValidatorAllocations) {
+        self.allocations.lock().unwrap().push(allocs)
     }
 }
 
 impl wasmtime_environ::Compiler for Compiler {
     fn compile_function(
         &self,
-        _translation: &ModuleTranslation<'_>,
-        _index: DefinedFuncIndex,
-        _data: FunctionBodyData<'_>,
-        _tunables: &Tunables,
+        translation: &ModuleTranslation<'_>,
+        index: DefinedFuncIndex,
+        data: FunctionBodyData<'_>,
+        tunables: &Tunables,
         _types: &ModuleTypes,
     ) -> Result<(WasmFunctionInfo, Box<dyn Any + Send>), CompileError> {
-        todo!()
+        let index = translation.module.func_index(index);
+        let sig = translation.get_types().function_at(index.as_u32()).unwrap();
+        let FunctionBodyData { body, validator } = data;
+        let start_srcloc = FilePos::new(
+            body.get_binary_reader()
+                .original_position()
+                .try_into()
+                .unwrap(),
+        );
+        let mut validator = validator.into_validator(self.take_allocations());
+        let env = FuncEnv::new(&translation.module, translation.get_types());
```

Comment:
Yeah I was thinking lazily initializing it on the fist compilation, and since functions are potentially compiled in parallel we'd need a mutex.

Reply:
I was taking a quick look at this, but I think it might be more complex than we want at this stage 🤔 The `FuncEnv` only works on references, so I don't think there are any extra allocations going on. Maybe it would be good to revisit if we need to store some data structures that are expensive to copy.

It's also a bit more complicated to save this since it operates on the module as a whole. We could potentially save it and use the module as a hash key, but I'm not sure we would know when to overwrite this structure the way `wasmtime` works today. Cranelift contexts seem to operating on a function which makes it a bit easier, since at the end of `compile_function` the context is no longer relevant so it can safely be overwritten on a subsequent run.